proxy/shadowsocks/spec_2022.go

```diff
@@ -0,0 +1,170 @@
+package shadowsocks
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"io"
+	"math/rand"
+	"sync/atomic"
+
+	"lukechampine.com/blake3"
+
+	"github.com/v2fly/v2ray-core/v5/common/buf"
+	"github.com/v2fly/v2ray-core/v5/common/crypto"
+	"github.com/v2fly/v2ray-core/v5/common/protocol"
+)
+
+const (
+	HeaderTypeClient      = 0
+	HeaderTypeServer      = 1
+	MaxPaddingLength      = 900
+	SaltSize              = 32
+	PacketNonceSize       = 24
+	MinRequestHeaderSize  = 1 + 8
+	MinResponseHeaderSize = MinRequestHeaderSize + SaltSize
+)
+
+var _ Cipher = (*AEAD2022Cipher)(nil)
+
+type AEAD2022Cipher struct {
+	KeyBytes           int32
+	AEADAuthCreator    func(key []byte) cipher.AEAD
+	UDPBlockCreator    func(key []byte) cipher.Block
+	UDPAEADAuthCreator func(key []byte) cipher.AEAD
+}
+
+func (c *AEAD2022Cipher) Family() CipherFamily {
+	if c.UDPBlockCreator != nil {
+		return CipherFamilyAEADSpec2022UDPBlock
+	} else {
+		return CipherFamilyAEADSpec2022
+	}
+}
+
+func (c *AEAD2022Cipher) KeySize() int32 {
+	return c.KeyBytes
+}
+
+func (c *AEAD2022Cipher) IVSize() int32 {
+	return SaltSize
+}
+
+func (c *AEAD2022Cipher) tcpAuthenticator(key []byte, iv []byte) *crypto.AEADAuthenticator {
+	subkey := make([]byte, c.KeyBytes)
+	deriveKey(key, iv, subkey)
+	aead := c.AEADAuthCreator(subkey)
+	nonce := crypto.GenerateAEADNonceWithSize(aead.NonceSize())
+	return &crypto.AEADAuthenticator{
+		AEAD:           aead,
+		NonceGenerator: nonce,
+	}
+}
+
+func (c *AEAD2022Cipher) NewEncryptionWriter(key []byte, iv []byte, writer io.Writer) (buf.Writer, error) {
+	auth := c.tcpAuthenticator(key, iv)
+	return crypto.NewAuthenticationWriter(auth, &crypto.AEADChunkSizeParser{
+		Auth: auth,
+	}, writer, protocol.TransferTypeStream, nil), nil
+}
+
+func (c *AEAD2022Cipher) NewDecryptionReader(key []byte, iv []byte, reader io.Reader) (buf.Reader, error) {
+	auth := c.tcpAuthenticator(key, iv)
+	return crypto.NewAuthenticationReader(auth, &crypto.AEADChunkSizeParser{
+		Auth: auth,
+	}, reader, protocol.TransferTypeStream, nil), nil
+}
+
+func (c *AEAD2022Cipher) EncodePacket(key []byte, b *buf.Buffer) error {
+	var nonceLen int32
+	payloadLen := b.Len()
+	var auth *crypto.AEADAuthenticator
+	if c.UDPBlockCreator != nil {
+		// aes
+		packetHeader := b.BytesTo(aes.BlockSize)
+		c.UDPBlockCreator(key).Encrypt(packetHeader, packetHeader)
```

Comment:
Block encryption should be done after AEAD sealing, because during AEAD we need the unencrypted session id to derive subkey and the packet id as nonce.